Support for ATEN iKVM servers

kelleyk opened the pull request:
Implements support for the 0x56 and 0x57 formats used by the AST2100-family (AST2400, etc.) BMCs, among others.

Closes #385.

kelleyk commented:
Hm.  Odd that the test suite should have problems; it works perfectly if you run `tests/run_from_console.js` locally.  From a glance at the output, seems like it might be an issue with how the tests are being run under Travis.

kelleyk commented:
Well, that was easy.  Was able to reproduce the failure locally with

`npm install karma`
`node_modules/karma/bin/karma start karma.conf.js`

and then quickly noticed that the `include/ast2100*.js` files weren't included in the list of source files in `karma.conf.js`.  Fixed!

jleroy commented:
Is this branch still a work in progress?
If not, is merging into master considered?

samhed commented:
I believe it's still work in progress, see #385 . Merging into master is definitely the plan!

And btw, @kelleyk it would be amazing if you guys could share the results of the reverse engineering done here to the rfbproto project: https://github.com/rfbproto/rfbproto

samhed commented:
@kelleyk is this something you will get back at working on?

kelleyk commented:
@samhed Is there anything in particular that you're asking about?

samhed commented:
I'm quite unsure of the state of this PR, for some reason I was under the impression that there were known problems that needed fixing. I haven't had the time to personally test the code.

phil21 commented:
@samhed Sorry I wasn't subscribed to this PR for some reason, so I didn't see your previous posts.  I'm chiming in because I was the wrangler of cats to get a bounty program setup - the results of which (and from others) is this pull request.  

At the time we submitted it we thought it was "ready" for a merge into master, but we only got silence.  I'm assuming it has to be rebased again at this point, but beyond that what exactly still needs work?  I think I speak for @kelleyk and @alantwentyseven when I say we're happy to help get this merged if there are any specific concerns.  I can allocate some dev resources from my team if small things need correcting as well.

It might be easiest to schedule an IRC/Slack chat with everyone at some point and just hammer out all the concerns plus get a punch list written down for what's still required to get this in.  It was definitely our intent to merge this into mainline moving forward, but we all kind of got burned out on it and never got much feedback.

kelleyk commented:
@phil21 did all of the hefty lifting in turning the raw technical work into the largely usable product it is today, that's for sure!  He also understands where I'm coming from well enough that he's welcome to speak for me.

I, at least, would very much appreciate either feel-good stories about people using the work we've done or specific criticism that we can address.

I can't say for sure, not having had the opportunity to chat with the upstream maintainers, but I suspect that any delay in taking this pull request is only because it is a relatively substantial set of changes.

samhed commented:
Thank you @kelleyk and @phil21! I apologize for not giving feedback earlier.

As you might have seen the code in master have changed a lot recently. Another rebase is definitely needed I'm afraid. I will try to be available to help out with this if you need. The commits are quite messy as well, it would be nice if this could be cleaned up while rebasing.

Other than adapting to the new code, I think we should put all the ast2100* files in a separate subdirectory under core/. And I would also like to see a bit of refactoring, to avoid cluttering the namespace, there are a lot of globals here.

The last, and perhaps biggest point for me is regarding the protocol. You have done some amazing work reverse-engineering here, but without a protocol the code will be difficult to maintain. Perhaps we can work together to get things documented and hopefully added to https://github.com/rfbproto/rfbproto

poptix commented:
This is a very important addition to noVNC for the company I work for. We participated in the bounty to implement it and are currently using it in production for thousands of servers. 

I'm very satisfied with the functionality and believe it to be a fine contribution to noVNC and the open source community as a whole. I look forward to seeing it merged with mainline.

samhed commented:
@poptix thank you for sharing! It helps knowing that the code is in production already. But it is still very difficult to keep maintained in the long run without a protocol specification. It doesn't have to be perfect, but something is better than just having to look at the code.

kelleyk commented:
@samhed Are you more interested in the ATEN-proprietary extensions to the RFB protocol, or in how the AST2100 video encoding works?

samhed commented:
> @samhed Are you more interested in the ATEN-proprietary extensions to the RFB protocol, or in how the AST2100 video encoding works?

Both I would assume, since the important part for me is to have some sort of protocol that supports the code to fall back on in the future.

kelleyk commented:
Okay, I'm getting a start on rebasing this branch on top of ef887cdb123df21b61043ff025e6208631e9eb7b.

There's quite a lot of work that mine is built on top of that was contributed by other people, so this might take a bit longer than it otherwise would.

Once I've gotten that squared away, I'll look into putting together some documentation.  I actually have a relatively extensive set of notes on this effort, but they're not all perfectly organized and I'll have to go through them to see what would be of interest to the `rfbproto` project.

samhed commented:
That sounds great @kelleyk!

kelleyk commented:
Okay, I've got it rebased, and have fixed a couple of obvious little mistakes that crept in.

My work is built on top of the pixel-format conversation work that @jimdigriz did, so I'm focusing on making my rebased version of his patches pass the test suite at the moment.

kelleyk commented:
I've got everything roughly rebased, but (probably as-to-be-expected given how much has changed upstream) there are some problems.  I have the client-side pixel format conversion patches passing their test suite, but there's still some work to be done on the others.

JonKohler commented:
hey @kelleyk - Were you ever able to make any forward progress on this last update?

We'd absolutely love to integrate this functionality into our company, where we'd be able to use this on 10s of thousands of Supermicro nodes, and would love to pick this support up off of master for long term supportability.

kelleyk commented:
I'm sorry for the very slow response time.  Yes, I took a good stab at cleaning up the patch series and rebasing it right after my last message, but then the pre-holiday rush hit at work and I had to put it down.

I've just gone through and effectively done that work again to rebase them on top of current `master`.  The maintainers have been making lots of long-overdue, good changes to noVNC itself; unfortunately, that makes rebasing both my patches and the client-side color conversion patches that they're built on a bit more tricky.  There have been some nontrivial changes to the UI, to the RFB implementation, etc.

At the moment, I have all of the tests passing, but the application itself cannot connect to iKVM servers yet.  I've fixed about a half-dozen small issues; at the moment, the connection flow works correctly up to the ClientInit phase of the RFB protocol.

I'm going to try to iron the last few little issues out, and then in parallel I'll clean up the patch series (there are many "oops, fix a little bug" commits that should just be squashed into the appropriate patch) and ask that folks test the result against whatever hardware they have handy.

If anyone is particularly antsy for access to my work-in-progress, you know where to find me!

kelleyk commented:
There are some imperfections with the rebased patch series:

- The old quality issue with IDCT (JPEG-ish) encoded blocks still exists.
- The UI elements need some love; the UI refactoring seems to have broken some of them.
- I'm not positive that I understand exactly how the JavaScript module system that noVNC now uses works; I would love for someone more familiar with it to take a look at how I've managed to get things to load and offer some guidance.

kelleyk commented:
Okay, I've updated my `bmc-support` branch.  Here is the updated patch series:

```
c5e4b79 2016-01-25 Alan Sharp OpenVNC (for iDRAC) fix; send only server-supported encodings
008f6a5 2016-01-28 Kevin Kelley Implement UI for controlilng ATEN-specific settings.
74f1e45 2016-01-14 Kevin Kelley Implement ATEN AST2100 (0x59) support.
2425612 2014-10-12 Alexander Clouter added aten-ikvm support
6b63b31 2014-10-12 Alexander Clouter added client side pixel format conversion support
```

CendioOssman commented:
You seem to have pushed to `bmc-support-rebase4` rather than `bmc-support`, so the new version isn't picked up here on this PR.

samhed commented:
> I'm not positive that I understand exactly how the JavaScript module system that noVNC now uses works; I would love for someone more familiar with it to take a look at how I've managed to get things to load and offer some guidance. (@samhed?)

Looks fine. Do you have any specific questions? Note that we have some out-commented code preparing for ES6 that might be confusing.

Overall, it's easier to comment on your code once it's synced with this PR. But i'll start with a few things:

* You seem to have removed the callback handler for 'onDesktopName' in app/ui.js
* Note that blue.css and black.css have been removed.
* Try to line break at 80 columns, at least where the surrounding code is trying to. I know our existing code isn't consistent at all here, but it's a good principle.
* Please remove any unrelated whitespace changes, there are quite a lot of them (mostly from Alex's commits though).

In general I would also like to see more split up commits and avoid these large lists in the commit messages. Would that be too much work?

kelleyk commented:
Yikes -- thanks @CendioOssman.  You can tell how sleep-deprived I was when I tried to push those patches!

Thanks for the quick feedback, @samhed!  I feel bad--I didn't mean to imply that the patch series was ready for review, just that I thought it was ready for folks to start practically experimenting with.

I've taken a pass at tidying up and resolving some small issues, including (I believe) the points that you brought up.  I've also tried to reflow comments and rearrange code a little bit to stay closer to 80 columns; let me know if there are any areas that still stick out to you.

I've removed Alan Sharp's Dell iDRAC patch because it seems like it might cause noVNC to ignore encodings that actually _are_ supported when connected to OpenVNC servers that _aren't_ Dell iDRACs.  However, it _is_ necessary for iDRAC support, from what I recall; perhaps we could reintroduce it (in another pull request?) with a configuration option that folks who know they're working with iDRACs can enable.

samhed commented on app/ui.js:
```diff
@@ -197,7 +197,7 @@ var UI;
             UI.initSetting('port', port);
             UI.initSetting('encrypt', (window.location.protocol === "https:"));
-            UI.initSetting('true_color', true);
+            UI.initSetting('convertColor', false);
```
Other settings such as view only uses underscore_separation instead of camelCase

samhed commented on core/rfb.js:
```diff
@@ -1437,14 +1468,97 @@
+                    pix = (pix << 8) | arr[k + offset];
+                } else {
+                    pix = (arr[k + offset] << (k*8)) | pix;
         }
```
Indentation

samhed commented on README.md:
```diff
@@ -140,7 +140,7 @@ WebSockets to TCP socket proxy. There is a python proxy included
     * Original Logo : Michael Sersen
     * tight encoding : Michael Tinglof (Mercuri.ca)
-    * pixel format conversion : [Alexander Clouter](http://www.digriz.org.uk/)
+    * pixel format conversion and ATEN iKVM support : [Alexander Clouter](http://www.digriz.org.uk/)
```
Two different lines imo

samhed commented on core/input/keysym.js:
```diff
@@ -379,4 +379,82 @@ var KeyTable = {
+
+// F{1..12}
+for (var i = KeyTable.XK_F1; i <= KeyTable.XK_F12; i++) {
+    XK2HID[i]			= 0x3a + (i - KeyTable.XK_F1);
```
4 spaces instead of tabs please

samhed commented on app/ui.js:
```diff
@@ -70,7 +70,8 @@ var UI;
                   "input/xtscancodes.js", "input/util.js", "input/devices.js",
-                  "display.js", "inflator.js", "rfb.js", "input/keysym.js"]});
+                  "display.js", "inflator.js", "rfb.js", "input/keysym.js",
+                  "ast2100/ast2100const.js", "ast2100/ast2100util.js", "ast2100/ast2100idct.js", "ast2100/ast2100.js"]});
```
line break after ast2100util.js

samhed commented on core/display.js:
```diff
@@ -672,12 +676,30 @@
-                img = this._drawCtx.createImageData(width, height);
-                img.data.set(new Uint8ClampedArray(arr.buffer, arr.byteOffset, width * height * 4));
-            }
+             if (SUPPORTS_UINT8CLAMPEDARRAY) {
```
Note the discussion in #581

samhed commented on core/rfb.js:
```diff
@@ -54,7 +57,13 @@
+        ['ATEN_ASTJPEG',        0x58 ],
+        ['ATEN_AST2100',        0x57 ],
+        ['ATEN_YARKON',         0x60 ],
+        ['ATEN_PILOT3',         0x61 ],
```
numerical order? :)

kelleyk commented:
Okay, I've addressed all of your comments except for the `SUPPORTS_UINT8CLAMPEDARRAY` one.  I believe that that is actually there because noVNC used to run its tests under PhantomJS 1.x, which did not support `UInt8ClampedArray` either.

It's on 2.x now, and from #581, it sounds like you aren't particularly worried about supporting old browsers, so I would be perfectly fine with removing it if you prefer.

samhed commented:
Thank you for your quick reponses!

> It's on 2.x now, and from #581, it sounds like you aren't particularly worried about supporting old browsers, so I would be perfectly fine with removing the flag if you prefer.

Sounds good to me.

As far as the commit size goes, I guess my main issue is with the "Add ATEN iKVM support."-commit which has a large list of changes in the commit-message. Would it be possible to split this commit into smaller commits? (And while we're at it, remove the periods after the commit headers :-P )

And lastly, as we have previously established, no one in the noVNC team has the possibility to test this code on a server. Before merging I would like to see:

* Some sort of commitment by you @kelleyk or someone else for trying to keep an eye out for related issues and maintain/support this code.
* An attempt at including the relevant parts of your notes into @rfbproto 
* Successful tests by someone with the hardware outside of @kelleyk's team. @JonKohler @poptix @jleroy would any of you be able to verify that Kevin's recent iteration of the bmc-support branch works?

CendioOssman commented on app/ui.js:
```diff
@@ -965,9 +983,171 @@ var UI;
+            var defaultSettings = {
+                quantTableSelectorLuma: defaultQuality,
+                quantTableSelectorChroma: defaultQuality,
+                subsamplingMode: parseInt(UI.ast2100_subsamplingMode)
```
TightVNC and TurboVNC also have these settings. Could we try to make a general UI and API rather than duplicating things?

kelleyk replied:
Do those things already exist in noVNC for the TIGHT encoding?  (`rfbproto` doesn't mention TurboVNC at all.) . I'm not familiar with the details of those other encodings, but if there's other functionality in noVNC that you feel this duplicates, I'd be happy to take a stab at combining them.

CendioOssman replied:
Currently these are hard coded. Probably because no-one has bothered adding a UI for it. :)

You have these settings:

https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst#jpeg-quality-level-pseudo-encoding
https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst#jpeg-fine-grained-quality-level-pseudo-encoding
https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst#jpeg-subsampling-level-pseudo-encoding

kelleyk replied:
This seems like it might be a little bit beyond the scope of this PR.

One would have to produce an interface that abstracts both the ATEN settings (chrominance and luminance quantization tables each on an independent scale from 0 to 11 and a choice of two subsampling modes) and the TIGHT settings (a single "quality level" and another single "fine quality level", plus the choice of several more subsampling levels); then one would have to "un-hard-wire" the TIGHT settings; provide a UI that can sit on top of both; etc.

CendioOssman replied:
We can wait with hooking up the Tight stuff (shouldn't be that hard though), but the API and UI should be generic from day one IMO.

CendioOssman commented on app/ui.js:
```diff
@@ -965,9 +983,171 @@ var UI;
+            } else {
+                Util.Info('Hiding AST2100 UI.');
+                propertyValue = 'none';
+            }
```
We've worked hard to clean up the code to separate visual things from logic. Please put stuff like this in the HTML and CSS and modify things by settings and removing CSS classes.

CendioOssman commented on app/ui.js:
```diff
@@ -965,9 +983,171 @@ var UI;
+                defaultValue = defaults.value,
+                snapToInt = true,
+                snapDuringDrag = true,
+                rect;
```
Why not a standard HTML element? A lot of work in order to construct your own widget.

kelleyk replied:
I cleaned it up and fixed a few issues with it, but the code was originally by Alan Sharp, if memory serves.  Still, that's a fair point; the `<input type="range">` widget would be a suitable choice.

CendioOssman commented on core/ast2100/ast2100.js:
```diff
@@ -0,0 +1,516 @@
+    // pick correct object to get the rfb websock from: in UI if enabled, in global rfb if using vnc_auto
+    var Xusock = typeof UI !== "undefined" ? UI.rfb._sock : rfb._sock;
+    RFB.messages.atenChangeVideoSettings(sock, lumaQt, chromaQt, subsamplingMode);
+};
```
This seems unnecessary and possibly even dangerous. Encapsulate things properly in the RFB object.

kelleyk replied:
We could even remove it; it was just a helper that I used to manually send those messages before a UI or configuration options existed.

CendioOssman commented on core/ast2100/ast2100.js:
```diff
@@ -0,0 +1,516 @@
+/*global ATEN_QT_LUMA, ATEN_QT_CHROMA, TABLE_CLASS_AC, TABLE_CLASS_DC */
+/*global BITS_AC_LUMA, BITS_AC_CHROMA, BITS_DC_LUMA, BITS_DC_CHROMA, HUFFVAL_AC_LUMA, HUFFVAL_AC_CHROMA, HUFFVAL_DC_LUMA, HUFFVAL_DC_CHROMA */
+/*global inRangeIncl, isEmpty */
+/*global fmt_u8a, fmt_u8, fmt_u16 */
```
There's a lot of namespace polution for these new files. Please try to sort that out so we can move towards being easy to integrate.

kelleyk replied:
I'd be happy to; do you have concrete suggestions?  Like I mentioned above, I'm not particularly familiar with noVNC's module system (which didn't exist when these patches were already written), and I don't work in JavaScript at the moment (professionally) so I'm a bit rusty on the finer points of project organization.

Are there choices other than moving everything defined in each file into a single object and then replacing references with e.g. `ast2100const.ATEN_QT_LUMA`, or was that what you had in mind?

CendioOssman replied:
I'm not a huge fan of JavaScript's scoping myself, so I have no really good answers except that your work might first need more module cleanup in the core files. :)

But until then you could try hiding everything under as few top level objects as possible. Ideally under the RFB object as it all is protocol stuff.

CendioOssman commented on core/ast2100/ast2100.js:
```diff
@@ -0,0 +1,516 @@
+        // Uint8Array(DCTSIZE2 * 4);
+        this._outputBuf = new Uint8Array(DCTSIZE2 * 4);
+        
+    };
```
All of this looks very much like JPEG. Any way we can leverage the browser's JPEG decoder?

kelleyk replied:
Unfortunately, no.  You're correct that some data is received in a format that looks like JPEG if you squint, but individual blocks that look much like JPEG MCUs can be interleaved with blocks compressed by vector quantization.

This should be covered in the cleaned-up version of my notes on the ins and outs of the ATEN RFB implementation and those on the 0x57 encoding.

CendioOssman commented on core/display.js:
```diff
@@ -470,9 +479,14 @@
+                if (isRgb) {
+                    this._rgbxImageData(x, y, width, height, arr, offset);
+                } else {
                 this._bgrxImageData(x, y, width, height, arr, offset);
```
Indent

CendioOssman commented on core/rfb.js:
```diff
@@ -148,16 +161,23 @@
         'shared': true,                         // Request shared mode
         'view_only': false,                     // Disable client mouse/keyboard
+        'aten_password_sep': ':',               // Separator for ATEN iKVM password fields
         'xvp_password_sep': '@',                // Separator for XVP password fields
```
Both of these are rather silly. We should change the API to include all necessary fields.

kelleyk replied:
Sure; I'll be happy to convert this to use explicit username/password fields as soon as they're available.

CendioOssman replied:
Feel free to add a commit that does this. I think it is just Xvp that user usernames right now. :)

kelleyk replied:
I agree that this would be a nice change, but it seems like it might be outside the scope of this PR.

CendioOssman replied:
I don't like piling on more stuff we don't really want. At minimum you should use the same separator.

CendioOssman commented on core/rfb.js:
```diff
@@ -919,6 +996,23 @@
+            }
+
+            // empty sub-auth list received means 'no auth' subtype selected
+            if (subAuthCount === 0) {
```
Why was this added now? Should probably be a separate commit.

kelleyk replied:
I'm not positive; this was part of @jimdigriz's original patch.

CendioOssman replied:
Please drop it in that case. If it is added I'd like to have a better rationale as why this wasn't needed until now.

kelleyk replied:
I don't know for a fact that it isn't necessary, and I'm nervous about fiddling with the detection heuristics and hand-shake process; getting that working properly was one of the more tedious parts of this project.

I suspect the rationale is something like "the ATEN iKVM servers do something weird and nonstandard", but let me fire up a debugger and investigate.

kelleyk replied:
I've removed this in the latest iteration on the patch series.

CendioOssman commented on core/rfb.js:
```diff
@@ -1069,53 +1172,141 @@
+                // keyboard portions of the test suite!
+                // 
+                RFB.messages.keyEvent = RFB.messages.atenKeyEvent;
+                RFB.messages.pointerEvent = RFB.messages.atenPointerEvent;
```
This is just obfuscation. Change the callers instead.

CendioOssman commented on core/rfb.js:
```diff
@@ -1069,53 +1172,141 @@
+            }
+
+            if (this._convert_color)
+                this._display.set_true_color(this._pixelFormat.true_color);
```
Won't the display always be true color after these changes? If so, just nuke everything regarding color maps from the display code.

CendioOssman commented on core/rfb.js:
```diff
@@ -1069,53 +1172,141 @@
+                        this._pixelFormat.depth = 8;
+                    }
+                    RFB.messages.pixelFormat(this._sock, this._pixelFormat);
             } else {
```
Indent

kelleyk replied:
Fixed, thanks!

CendioOssman commented on core/rfb.js:
```diff
@@ -1271,6 +1462,42 @@
             }
 
+            if (this._rfb_atenikvm) {
+                switch (msg_type) {
```
Do these conflict with existing message?

They should be in the main switch. And in rfbproto. And reported to IANA.

kelleyk replied:
Yes, some do conflict; for example, per https://github.com/kelleyk/rfbproto/blob/master/rfbproto.rst#server-to-client-messages, 4 should be a "ResizeFrameBuffer" event.  Regardless, they're proprietary ATEN extensions, and it seems like a good idea to keep them a little bit separate.

CendioOssman replied:
Argh. Well that's horribly annoying. Let's hope they never conflict in a way that forces us to choose. :/

Please put a small comment here though explaining that they conflict.

kelleyk replied:
Okay, done!

CendioOssman commented on core/rfb.js:
```diff
@@ -1271,6 +1462,42 @@
+                    case 60:  // Get Viewer Lang
+                        Util.Debug("ATEN iKVM Get Viewer Lang");
+                        this._sock.rQskipBytes(8);
+                        return true;
```
Doesn't any of these have to report back to the server?

kelleyk replied:
The server doesn't seem to mind.

CendioOssman commented on core/rfb.js:
```diff
@@ -1486,6 +1810,11 @@
+    // ATEN iKVM doesn't send the resolution until the first refresh, so start big
+    // to avoid a repaint
+    RFB.ATEN_INIT_WIDTH = 10000;
+    RFB.ATEN_INIT_HEIGHT = 10000;
```
These only seem to be used in one place, so they don't need constants.

CendioOssman commented on core/rfb.js:
```diff
@@ -1543,6 +1872,90 @@
         },
 
+        atenKeyEvent: function (sock, keysym, down) {
+            var ks = XK2HID[keysym];
```
This looks like it assumes a US keyboard layout. You should look at the Qemu code on how to do this properly.

kelleyk replied:
It maps keysyms to the USB HID scancodes that the ATEN iKVM expects.  I believe (but am not 100% positive) that the virtual USB device presented to the host by the iKVM server always reports itself as having a US keyboard layout.

I'm not sure what alternative would be better.   We could send whatever scancode the browser gives us to the iKVM server and hope that the host OS interprets it correctly, but that seems dangerous.

I don't have non-US-layout keyboards or non-US-locale servers to test this with, unfortunately.

CendioOssman replied:
> I believe (but am not 100% positive) that the virtual USB device presented to the host by the iKVM server always reports itself as having a US keyboard layout.

All keyboards tend to do that because manufacturers are lazy. Hence no OS respects that information. :)

> I'm not sure what alternative would be better. We could send whatever scancode the browser gives us to the iKVM server and hope that the host OS interprets it correctly, but that seems dangerous.

That would be the correct approach. Otherwise noVNC would need to have a lookup table per keyboard layout.

> I don't have non-US-layout keyboards or non-US-locale servers to test this with, unfortunately.

What's printed on the keys doesn't matter. Just configure your local settings to something different. French is a favourite in finding bugs. :)

kelleyk replied:
I took a stab at this and found a further complication: what the browser supplies[1] is *not* a USB HID usage code ("scan-code")[2].

[1] https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key
[2] http://www.usb.org/developers/hidpage/Hut1_12v2.pdf, ch. 10 (page 53)

CendioOssman replied:
Right, the browsers have their own systems. Look at xtscancode.js for mapping between that and XT codes. We should probably merge that table with yours.

CendioOssman commented on core/rfb.js:
```diff
@@ -1543,6 +1872,90 @@
+            sock._sQlen += 18;
+        },
+
+        atenPointerEvent: function (sock, x, y, mask) {
```
Any idea why they didn't use the standard message for this?

kelleyk replied:
No idea, offhand.  They don't seem to have prioritized sticking to available standard stuff.

CendioOssman replied:
That makes sense for the keyboard event, but not really mouse events. It's just a simple coordinate either way.

PS. What happens if you send a normal keyboard event?

kelleyk replied:
I can't say for sure until I locate my notes, but I'm fairly confident that it would be ignored.  I don't recall seeing anything related to multiple types of key or pointer events in the iKVM server.

CendioOssman commented on vnc.html:
```diff
@@ -200,7 +200,7 @@ <h1 class="noVNC_logo" translate="no"><span>no</span><br />VNC</h1>
                         <div><ul>
                             <li>
-                                <label><input id="noVNC_setting_true_color" type="checkbox" checked /> True Color</label>
+                                <label><input id="noVNC_setting_convert_color" type="checkbox" checked /> Convert Color</label>
```
I'm not sure this makes sense. True Color is broken right now anyway, so I say keep this stuff as an internal thing which is only used for broken servers like the ATEN ones.

kelleyk replied:
To clarify: You'd like me to remove the UI element and have this mode automatically enabled only for e.g. ATEN servers?

CendioOssman replied:
Yup.

kelleyk replied:
Okay, done!

kelleyk commented:
In response to @samhed 's comment:

I'll take another look at the "Add ATEN iKVM support" commit.  I'm not the original author, so I just left it the way it was.

As far as the commit message: The list of changes are not changes made by the commit; they describe changes made between @jimdigriz's original patch and the version I'm submitting here.  I'll take a glance at that, too, but it's important to me to give credit where credit is due, and the commit message seems like the appropriate place to describe the provenance of a patch.

Re: a commitment -- hopefully the fact that I've put so much time into #385 and this PR speaks for itself; but sure, I'll be happy to continue to maintain/support this functionality as time allows.

Re: my notes -- I need to spend a few minutes figuring out where I've stuck everything, but yes, I still intend to do this.

samhed commented:
Nice! Thank you @kelleyk

CendioOssman commented:
> As far as the commit message: The list of changes are not changes made by the commit; they describe changes made between @jimdigriz's original patch and the version I'm submitting here. I'll take a glance at that, too, but it's important to me to give credit where credit is due, and the commit message seems like the appropriate place to describe the provenance of a patch.

Git has the `--author` flag you can use if some parts are mostly someone else's.

However, I strongly second @samhed's request for splitting this up. It's very difficult to review such a large commit.

CendioOssman commented:
@kelleyk, in case you didn't notice I added #756 as an idea that might help you out with modules and scoping. I'd appreciate if you could have a look and see how it fits for your needs.

kelleyk commented:
Sorry for the slow reply; yes, I think it would.  I suppose I'll hang on for a little while and see how the conversation in that issue proceeds.

kelleyk commented:
Which issues that would prevent landing this work are still outstanding, from your perspectives?

The ones that seem most pressing to me are
- improving the modularity of the new code, an effort which is blocked on #756; and
- the keyboard issue, which is blocked on an understanding of what the most "honest" way to pass USB HID usage codes to the iKVM server is.

Does that match your understanding?

I don't mean to exclude the other things that you've helpfully pointed out; I'm happy to make those changes, too, but I'd like to make sure we're on the same page as far as the serious issues go.

Thank you for all of the effort you've already put towards getting this merged!

CendioOssman commented:
> The ones that seem most pressing to me are
> 
>   * improving the modularity of the new code, an effort which is blocked on #756; and
>   * the keyboard issue, which is blocked on an understanding of what the most "honest" way to pass USB HID usage codes to the iKVM server is.
> 
> Does that match your understanding?

Right. And I'd like to again stress the need for splitting up the commits. We've most likely missed things trying to review this huge diff.

For the keyboard, have a look at #766 as well. You should produce a mapping from `KeyboardEvent.code` to USB HID codes.

CendioOssman commented:
The modularisation of noVNC is now merged, so you could have a look at rebaseing this.

pasikarkkainen commented:
Are there plans to rebase this PR ? Thanks.

pasikarkkainen commented:
@kelleyk is it likely you'll be able to rebase / rework this PR ? Thanks.

CendioOssman commented:
Closing this as @kelleyk seems to have gone missing.

kelleyk commented:
Howdy--I'm sorry, I haven't gone missing; I've just been busy enough with work that I haven't had time to rework this stuff to fit into the new module system.

CendioOssman commented:
Alright. We'll be here when you are ready to resume this work. :)

AndrewJDR commented:
We're using https://github.com/kelleyk/noVNC in production to connect to several servers. Would love to see this merged.
